### PR TITLE
[proxima-core] #137 enforce using Repository#asFactory when serializing Repository

### PR DESCRIPTION
**Pull request by je-ik**

### CODE REVIEW

File: beam/core/src/test/java/cz/o2/proxima/beam/core/AttributeFamilyProxyDataDescriptorTest.java [hunk 1]
```diff
@@ -49,7 +49,6 @@ public void setUp() {
   public void tearDown() {
     direct.close();
     beam.close();
-    repo.discard();
```
Comment (je-ik) [inline beam/core/src/test/java/cz/o2/proxima/beam/core/AttributeFamilyProxyDataDescriptorTest.java]: The purpose of this PR is mostly to get rid of these ugly calls.

**Comment by je-ik** [comment]
This closes O2-Czech-Republic#137

File: core/src/main/java/cz/o2/proxima/repository/ConfigRepository.java [hunk 2]
```diff
@@ -1957,4 +1972,43 @@ public boolean isEmpty() {
...
+  }
+
+  private void readObject(ObjectInputStream ois) throws IOException, ClassNotFoundException {
+    RepositoryFactory readFactory = (RepositoryFactory) ois.readObject();
+    this.factory = readFactory;
+  }
+
+  private Object readResolve() throws ObjectStreamException {
+    Preconditions.checkState(this.factory != null);
+    return this.factory.apply();
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/repository/ConfigRepository.java]: 👍

File: core/src/main/java/cz/o2/proxima/repository/ConfigRepository.java [hunk 3]
```diff
@@ -58,7 +63,7 @@
 @Slf4j
 // BUG in error prone
 @SuppressWarnings("InconsistentCapitalization")
-public class ConfigRepository extends Repository {
+public final class ConfigRepository extends Repository implements Serializable {
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/repository/ConfigRepository.java]: Repository already implements Serializable

File: core/src/main/java/cz/o2/proxima/repository/Repository.java [hunk 4]
```diff
@@ -73,8 +65,31 @@ public static Repository of(Config config) {
     return ConfigRepository.of(config);
   }
 
+  public interface ConfigFactory extends Serializable {
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/repository/Repository.java]: javadoc

File: core/src/main/java/cz/o2/proxima/repository/RepositoryFactory.java [hunk 5]
```diff
@@ -28,8 +31,12 @@
 
     private final RepositoryFactory underlying;
 
-    private Caching(RepositoryFactory underlying) {
+    private Caching(RepositoryFactory underlying, Repository created) {
       this.underlying = underlying;
+      synchronized (Repository.class) {
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/repository/RepositoryFactory.java]: Looking at this class, I don't really think that any of the fields should be in static context. What if you have multiple repositories per JVM? This doesn't look right..

**Comment by dmvk** [comment]
Hmm, I think I see what you meant by the "caching" repository - having a JVMwide cache for beam pipelines. I that case we should:

- Introduce `Repository#getInstanceId` method
- Generate a new UUID when the repository is first created (on driver side) and store it in `instanceId` field
- Having a static "map" based cache for multiple instances (`Map<String /* InstanceId */, Repossitory>`
- This way we can also merge `caching` and `local` repository, as getInstanceId would efectively replace identityHashCode

WDYT?

**Comment by je-ik** [comment]
The assumption, that Repository is a singleton is present throughout the codebase. Allowing multiple repositories within single JVM would require much deeper changes - from the top of my head, attribute and entity descriptors are not linked to Repository that created them, and so on. Before we dive into any changes that would allow multiple repositories, I'd like to know *why* would anyone want that (with the exception of testing which what this PR is trying to systematically solve). I pretty much think it would cause more harm than good. I think we can enforce this restriction a little more - we can check that we don't create more Repository objects (without test flag) in single JVM in constructor of Repository

**Comment by dmvk** [comment]
👍 for introducing the check, if that would be as strict as failing the second call to `Repository.of(...)` per JVM

Maybe in that case we should also introduce `Repository.getInstance()` method?

**Comment by dmvk** [comment]
Second thoughts ... `Repository.getInstance()`, may tie our hands in the future...

**Comment by je-ik** [comment]
Agree, I would want to let users access `Repository` via static context.

**Comment by je-ik** [comment]
@dmvk see [this commit](https://github.com/datadrivencz/proxima-platform/pull/294/commits/acbb075211e2bb02a17af0ed961e810026a0b1e5)

File: core/src/main/java/cz/o2/proxima/repository/ConfigRepository.java [hunk 6]
```diff
@@ -293,6 +298,23 @@ private ConfigRepository(
     } catch (Exception ex) {
       throw new IllegalArgumentException("Cannot read config settings", ex);
     }
+
+    if (cachingEnabled) {
+      if (cachedConfigConstructed != null) {
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/repository/ConfigRepository.java]: is this necessary?

Comment (dmvk) [reply core/src/main/java/cz/o2/proxima/repository/ConfigRepository.java]: I mean, shouldn't we always fail fast?

**Pull request merged**